kliff/transforms/configuration_transforms/descriptors.py

```diff
@@ -0,0 +1,469 @@
+        raise DescriptorsError("Libdescriptor not available.")
+    print("-" * 80)
+    print(
+        "Descriptors below are currently available, select them by `descriptor: str` attribute:"
+    )
+    print("-" * 80)
+    _instance = AvailableDescriptors()
+    for key in _instance.__dict__.keys():
+        print(f"{key}")
+
+
+@requires(libdescriptor_available, "libdescriptor is needed for Descriptors")
+class Descriptor(ConfigurationTransform):
+    """
+    Descriptor class provides interface with the libdescriptor library. It provides a
+    unified interface to all the descriptors in libdescriptor, however all descriptors need
+    to have a corresponding initializer routine, to deal with hyperparameters. The descriptor
+    is initialized with a cutoff radius, a list of species,a descriptor type and an ordered
+    list of hyperparameters. The descriptor type is a string, which can be obtained by
+    `show_available_descriptors()` function. Some sane default values for the hyperparameters
+    are provided in ~:kliff.transforms.configuration_transforms.default_hyperparameters
+    module. This class methods to compute derivatives of the descriptor with respect to
+    atomic positions. The functions generating descriptors and their derivatives are
+    implemented as `forward()` and `backward()`, respectively, to match the PyTorch nomenclature.
+
+    """
+
+    def __init__(
+        self,
+        cutoff: float,
+        species: List[str],
+        descriptor: str,
+        hyperparameters: Union[Dict, str],
+        cutoff_function: str = "cos",
+        nl_ctx: NeighborList = None,
+        implicit_fingerprint_copying: bool = False,
+    ):
+        """
+        Args:
+            cutoff (float): Cutoff radius.
+            species (list): List of strings, each string is a species (atomic symbols).
+            descriptor (str): String of descriptor type, can be obtained by `show_available_descriptors()`.
+            hyperparameters (OrderedDict): Ordered dictionary of hyperparameters.
+            cutoff_function (str): Cut-off function, currently only "cos" is supported.
+            nl_ctx (NeighborList): function to compute neighbor list, if not provided, will be computed internally.
+            implicit_fingerprint_copying (bool): If True, the fingerprint will be copied to
+                the Configuration object's fingerprint attribute.
+        """
+        super().__init__(implicit_fingerprint_copying)
+        self.cutoff = cutoff
+        self.species = species
+        _available_descriptors = AvailableDescriptors()
+        self.descriptor_name = descriptor
+        self.descriptor_kind = getattr(_available_descriptors, descriptor)
+        self.width = -1
+        self.hyperparameters = self._set_hyperparams(hyperparameters)
+        self.cutoff_function = cutoff_function
+        self._cdesc, self.width = self._init_descriptor_from_kind()
+        if nl_ctx:
+            self.nl_ctx = nl_ctx
```

Comment:
In the forward and backward functions below, a neighbor list should be created for each config. Here, if a user provides `nl_ctx`, can it be used later in the two functions?

Reply:
This comment is unclear, isn't that what is happening? if user provides external neigbor list context then it is used in forward and backward function. Otherwise it will compute own neigbor lists. But I think it might not be that good of a design. It was supposed to work with other neigbor lists and reusing precomputed neighborlists. With implicit fingerprint copying I think I can remove this one.

Reply:
Does that answer your question?

Reply:
I leave a comment. See that.

Reply:
Replied to previous comment.